libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -597,6 +601,186 @@ class plugin final : public virtual operator_plugin<deduplicate_operator> {
+        // we reset the entry.
+        if (cfg_.create_timeout_
+            and now > it->second.created_at + cfg_.create_timeout_->inner) {
+          it.value().reset(current_row, now);
+          ids.append_bit(true);
+          continue;
+        }
+        if (cfg_.write_timeout_
+            and now > it->second.written_at + cfg_.write_timeout_->inner) {
+          it.value().reset(current_row, now);
+          ids.append_bit(true);
+          continue;
+        }
+        if (cfg_.read_timeout_
+            and now > it->second.read_at + cfg_.read_timeout_->inner) {
+          it.value().reset(current_row, now);
+          ids.append_bit(true);
+          continue;
+        }
+        // If there's a distance configured, and the distance has elapsed, we
+        // erase the entry.
+        if (cfg_.distance_
+            and current_row > it->second.last_row + cfg_.distance_->inner) {
+          it.value().reset(current_row, now);
+          ids.append_bit(true);
+          continue;
+        }
+        it.value().read_at = now;
+        it.value().last_row = current_row;
+        if (not cfg_.limit_) {
+          ids.append_bit(false);
+          continue;
+        }
+        if (it->second.count >= cfg_.limit_->inner) {
+          ids.append_bit(false);
+          continue;
+        }
+        it.value().count += 1;
+        it.value().written_at = now;
+        ids.append_bit(true);
+      }
+      row += keys.length();
+      for (auto [begin, end] : select_runs(ids)) {
+        co_yield subslice(events, begin, end);
+      }
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.deduplicate";
+  }
+
+  auto optimize(const expression& filter, event_order order) const
+    -> optimize_result override {
+    (void)order;
+    if (cfg_.distance_) {
+      // When the `distance` option is used, we're not allowed to optimize at
+      // all. Here's a simple example that proves this:
+      //   metrics "platform"
+      //   deduplicate connected, distance=1
+      //   where not connected
+      return do_not_optimize(*this);
+    }
+    return optimize_result{filter, event_order::ordered, copy()};
+  }
+
+  friend auto inspect(auto& f, deduplicate_operator2& x) -> bool {
+    return f.object(x).fields(f.field("configuration", x.cfg_));
+  }
+
+private:
+  configuration2 cfg_{};
+};
+
+class tql2_plugin final
+  : public virtual operator_plugin2<deduplicate_operator2> {
+public:
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto cfg = configuration2{};
+    auto parser = argument_parser2::operator_("deduplicate");
+    parser.positional("expr", cfg.expr_, "<expr>");
+    parser.named("distance", cfg.distance_);
+    parser.named("limit", cfg.limit_);
+    parser.named("create_timeout", cfg.create_timeout_);
+    parser.named("write_timeout", cfg.write_timeout_);
+    parser.named("read_timeout", cfg.read_timeout_);
```

Comment:
I still need to write docs, but before that I wanted to hear what @mavam thinks about separating `timeout` into three options. The TQL1 operator only had the equivalent of `read_timeout`.

The timeouts behave the following way:
- `create_timeout` is determined at the time when the first event for a key is let through
- `write_timeout` is determined whenever an event for a key is let through; if `limit` is 1, this is the same as `create_timeout`
- `read_timeout` is determined whenever a key is found, no matter whether an event was let through or not.

This allows saying things like `deduplicate foo, limit=10, create_timeout=1h` to let 10 events per unique `foo` through every hour, which was not possible with the TQL1 operator.